src/daemon/auth/common/crypt.rs

```diff
@@ -38,10 +38,61 @@ const POLY1305_TAG_BYTE_LEN: usize = POLY1305_TAG_BIT_LEN / 8;
 const CLEARTEXT_PREFIX_LEN: usize = CHACHA20_NONCE_BYTE_LEN + POLY1305_TAG_BYTE_LEN;
 const UNUSED_AAD: [u8; 0] = [0; 0];
 
+//------------ CrossPlatformDeserializableAtomicU64 -------------------------------------------------------------------
+
+/// This type provides Serde (de)serialization irrespective of target platform, as the Serde crate provided derive
+/// impls only appear to produce code when compiled on x86_64. See: https://github.com/NLnetLabs/krill/issues/751
+#[derive(Debug)]
+struct CrossPlatformDeserializableAtomicU64(AtomicU64);
+
+impl std::ops::Deref for CrossPlatformDeserializableAtomicU64 {
+    type Target = AtomicU64;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for CrossPlatformDeserializableAtomicU64 {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct DeserializableAtomicU64Visitor;
+
+        impl<'de> serde::de::Visitor<'de> for DeserializableAtomicU64Visitor {
+            type Value = CrossPlatformDeserializableAtomicU64;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("an unsigned integer in the range 0..=2^64-1")
+            }
+
+            fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error, {
+                Ok(CrossPlatformDeserializableAtomicU64(AtomicU64::new(v)))
+            }
+        }
+
+        deserializer.deserialize_u64(DeserializableAtomicU64Visitor)
+    }
+}
+
+impl serde::Serialize for CrossPlatformDeserializableAtomicU64 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_u64(self.0.load(Ordering::SeqCst))
+    }
+}
+
+//------------ NonceState ---------------------------------------------------------------------------------------------
+
 #[derive(Debug, Deserialize, Serialize)]
 pub struct NonceState {
-    sender_unique: [u8; 4], //   32 bits
-    counter: AtomicU64,     // + 64 bits = 96 bits = CHACHA20_NONCE_BIT_LEN
+    sender_unique: [u8; 4],           //   32 bits
+    counter: CrossPlatformDeserializableAtomicU64, // + 64 bits = 96 bits = CHACHA20_NONCE_BIT_LEN
```

Comment:
I realise this was not introduced now, but I wonder if it would be better to use 64 bits for `sender_unique` and 32 bits (4 billion) for `counter`. Although the chance of a collision for `sender_unique` is small with 32 bits, it's much smaller still with 64 - and 4 billion feels like enough space for the counter. But, just a thought - I am not 100% sure what is best.

Reply:
This design was based on [RFC 8439 Section 2.3](https://datatracker.ietf.org/doc/html/rfc8439#section-2.3) which says:
> In cases where a single key is used by multiple senders, it is important to make sure that they don't use the same nonces.  This can be assured by partitioning the nonce space so that the first 32 bits are unique per sender, while the other 64 bits come from a counter.

Reply:
ok sure.. I think 32 bits is probably enough to not have any collisions (since they all start with counter 0)